bigquery/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryMetaStoreClientImpl.java

```diff
@@ -0,0 +1,620 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.apache.iceberg.BaseMetastoreTableOperations;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.exceptions.BadRequestException;
+import org.apache.iceberg.exceptions.ForbiddenException;
+import org.apache.iceberg.exceptions.NoSuchIcebergTableException;
+import org.apache.iceberg.exceptions.NoSuchNamespaceException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.exceptions.NotAuthorizedException;
+import org.apache.iceberg.exceptions.RuntimeIOException;
+import org.apache.iceberg.exceptions.ServiceFailureException;
+import org.apache.iceberg.exceptions.ServiceUnavailableException;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+/** A client of Google Bigquery Metastore functions over the BigQuery service. */
+public final class BigQueryMetaStoreClientImpl implements BigQueryMetaStoreClient {
+
+  public static final ExceptionHandler.Interceptor EXCEPTION_HANDLER_INTERCEPTOR =
+      new ExceptionHandler.Interceptor() {
+
+        private static final long serialVersionUID = -8429573486870467828L;
+
+        @Override
+        public RetryResult afterEval(Exception exception, RetryResult retryResult) {
+          return ExceptionHandler.Interceptor.RetryResult.CONTINUE_EVALUATION;
+        }
+
+        @Override
+        public RetryResult beforeEval(Exception exception) {
+          if (exception instanceof BaseServiceException) {
+            boolean retriable = ((BaseServiceException) exception).isRetryable();
+            return retriable
+                ? ExceptionHandler.Interceptor.RetryResult.RETRY
+                : ExceptionHandler.Interceptor.RetryResult.CONTINUE_EVALUATION;
+          }
+          return ExceptionHandler.Interceptor.RetryResult.CONTINUE_EVALUATION;
+        }
+      };
+
+  private final Bigquery client;
+  private final BigQueryOptions bigqueryOptions;
+
+  // Retry config with error messages and regex for rate limit exceeded errors.
+  private static final BigQueryRetryConfig DEFAULT_RETRY_CONFIG =
+      BigQueryRetryConfig.newBuilder()
+          .retryOnMessage(BigQueryErrorMessages.RATE_LIMIT_EXCEEDED_MSG)
+          .retryOnMessage(BigQueryErrorMessages.JOB_RATE_LIMIT_EXCEEDED_MSG)
+          .retryOnRegEx(BigQueryErrorMessages.RetryRegExPatterns.RATE_LIMIT_EXCEEDED_REGEX)
+          .build();
+
+  public static final ExceptionHandler BIGQUERY_EXCEPTION_HANDLER =
+      ExceptionHandler.newBuilder()
+          .abortOn(RuntimeException.class)
+          .retryOn(java.net.ConnectException.class) // retry on Connection Exception
+          .retryOn(java.net.UnknownHostException.class) // retry on UnknownHostException
```

Comment:
"retry on UnknownHostException" is obvious when reading the code. It would be nice to explain **_why_** we retry on such exceptions.